bin/uniffle-function.sh

```diff
@@ -0,0 +1,376 @@
+  declare array=("${!arrref}")
+  declare oifs
+
+  declare globstatus
+  declare -a sa
+
+  globstatus=$(set -o | grep noglob | awk '{print $NF}')
+
+  set -f
+  oifs=${IFS}
+
+  IFS=$'\n' sa=($(sort <<<"${array[*]}"))
+
+  eval "${arrname}"=\(\"\${sa[@]}\"\)
+
+  IFS=${oifs}
+  if [[ "${globstatus}" = off ]]; then
+    set +f
+  fi
+}
+
+#---
+# uniffle_generate_usage: generate standard usage output and optionally takes a class.
+# @param1 execname
+# @param2 true|false
+# @param3 [text to use in place of SUBCOMMAND]
+#---
+function uniffle_generate_usage
+{
+  declare cmd=$1
+  declare takesclass=$2
+  declare subcmdtext=${3:-"SUBCOMMAND"}
+  declare haveoptions
+  declare optstring
+  declare havesubs
+  declare subcmdstring
+  declare cmdtype
+
+  cmd=${cmd##*/}
+
+  if [[ -n "${UNIFFLE_OPTION_USAGE_COUNTER}"
+        && "${UNIFFLE_OPTION_USAGE_COUNTER}" -gt 0 ]]; then
+    haveoptions=true
+    optstring=" [OPTIONS]"
+  fi
+
+  if [[ -n "${UNIFFLE_SUBCMD_USAGE_COUNTER}"
+        && "${UNIFFLE_SUBCMD_USAGE_COUNTER}" -gt 0 ]]; then
+    havesubs=true
+    subcmdstring=" ${subcmdtext} [${subcmdtext} OPTIONS]"
+  fi
+
+  echo "Usage: ${cmd}${optstring}${subcmdstring}"
+  if [[ ${takesclass} = true ]]; then
+    echo " or    ${cmd}${optstring} CLASSNAME [CLASSNAME OPTIONS]"
+    echo "  where CLASSNAME is a user-provided Java class"
+  fi
+
+  if [[ "${haveoptions}" = true ]]; then
+    echo ""
+    echo "  OPTIONS is none or any of:"
+    echo ""
+
+    uniffle_generic_columnprinter "" "${UNIFFLE_OPTION_USAGE[@]}"
+  fi
+
+  if [[ "${havesubs}" = true ]]; then
+    echo ""
+    echo "  ${subcmdtext} is one of:"
+    echo ""
+
+    if [[ "${#UNIFFLE_SUBCMD_USAGE_TYPES[@]}" -gt 0 ]]; then
+      uniffle_sort_array UNIFFLE_SUBCMD_USAGE_TYPES
+      for subtype in "${UNIFFLE_SUBCMD_USAGE_TYPES[@]}"; do
+        cmdtype="$(tr '[:lower:]' '[:upper:]' <<< ${subtype:0:1})${subtype:1}"
+        printf "\n    %s Commands:\n\n" "${cmdtype}"
+        uniffle_generic_columnprinter "${subtype}" "${UNIFFLE_SUBCMD_USAGE[@]}"
+      done
+    else
+      uniffle_generic_columnprinter "" "${UNIFFLE_SUBCMD_USAGE[@]}"
+    fi
+    echo ""
+    echo "${subcmdtext} may print help when invoked w/o parameters or with -h."
+  fi
+}
+
+#---
+# uniffle_generic_columnprinter: Print a screen-size aware two-column output,
+# if reqtype is not null, only print those requested.
+# @param1 reqtype
+# @param2 array
+#---
+function uniffle_generic_columnprinter
+{
+  declare reqtype=$1
+  shift
+  declare -a input=("$@")
+  declare -i i=0
+  declare -i counter=0
+  declare line
+  declare text
+  declare option
+  declare giventext
+  declare -i maxoptsize
+  declare -i foldsize
+  declare -a tmpa
+  declare numcols
+  declare brup
+
+  if [[ -n "${COLUMNS}" ]]; then
+    numcols=${COLUMNS}
+  else
+    numcols=$(tput cols) 2>/dev/null
+    COLUMNS=${numcols}
+  fi
+
+  if [[ -z "${numcols}"
+     || ! "${numcols}" =~ ^[0-9]+$ ]]; then
+     numcols=75
```

Comment:
Why set `numcols` as 75?

Reply:
In command-line interfaces (CLI) , fixed widths of around `80` or `120` are commonly used for displaying output results. The value 75 is an empirical value.

Reply:
Thanks for your explanation. I see.